src/pessimisticPatternMatcherUtils.cc

```diff
@@ -0,0 +1,115 @@
+                uint16_t ref_ctr_id, ndarray::Array<double, 1, 1> const& proj_ref_ctr_delta,
+                double proj_ref_ctr_dist_sq, ndarray::Array<long int, 1, 1> const& ref_dist_idx_array,
+                ndarray::Array<uint16_t, 1, 1> const& ref_id_array,
+                ndarray::Array<double, 2, 1> const& reference_array, double src_sin_tol) {
+    /// Loop over our candidate reference objects.
+    for (auto idx = ref_dist_idx_array.begin(); idx != ref_dist_idx_array.end(); idx++) {
+        /// Compute the delta vector from the pattern center.
+        unsigned int ref_id = ref_id_array[*idx];
+        double ref_delta[3];
+        double proj_ref_delta[3];
+        ref_delta[0] = reference_array[ref_id][0] - ref_ctr[0];
+        ref_delta[1] = reference_array[ref_id][1] - ref_ctr[1];
+        ref_delta[2] = reference_array[ref_id][2] - ref_ctr[2];
+
+        double ref_dot = ref_delta[0] * ref_ctr[0] + ref_delta[1] * ref_ctr[1] + ref_delta[2] * ref_ctr[2];
+        proj_ref_delta[0] = ref_delta[0] - ref_dot * ref_ctr[0];
+        proj_ref_delta[1] = ref_delta[1] - ref_dot * ref_ctr[1];
+        proj_ref_delta[2] = ref_delta[2] - ref_dot * ref_ctr[2];
+        /// Compute the cos between our "center" reference vector and the
+        /// current reference candidate.
+        double proj_delta_dist_sq =
+                (proj_ref_delta[0] * proj_ref_delta[0] + proj_ref_delta[1] * proj_ref_delta[1] +
+                 proj_ref_delta[2] * proj_ref_delta[2]);
+        double geom_dist_ref = sqrt(proj_ref_ctr_dist_sq * proj_delta_dist_sq);
+        double cos_theta_ref =
+                (proj_ref_delta[0] * proj_ref_ctr_delta[0] + proj_ref_delta[1] * proj_ref_ctr_delta[1] +
+                 proj_ref_delta[2] * proj_ref_ctr_delta[2]) /
+                geom_dist_ref;
+
+        /// Make sure we can safely make the comparison in case
+        /// our "center" and candidate vectors are mostly aligned.
+        double cos_sq_comparison;
+        if (cos_theta_ref * cos_theta_ref < (1 - src_sin_tol * src_sin_tol)) {
+            cos_sq_comparison = (cos_theta_src - cos_theta_ref) * (cos_theta_src - cos_theta_ref) /
+                                (1 - cos_theta_ref * cos_theta_ref);
+        } else {
+            cos_sq_comparison = (cos_theta_src - cos_theta_ref) * (cos_theta_src - cos_theta_ref) /
+                                (src_sin_tol * src_sin_tol);
+        }
+        /// Test the difference of the cosine of the reference angle against
+        /// the source angle. Assumes that the delta between the two is
+        /// small.
+        if (cos_sq_comparison > src_sin_tol * src_sin_tol) {
+            continue;
+        }
+        /// The cosine tests the magnitude of the angle but not
+        /// its direction. To do that we need to know the sine as well.
+        /// This cross product calculation does that.
+        double cross_ref[3];
+        cross_ref[0] =
```

Comment:
You should be able to do all this with Eigen, but I'm not positive about the exact syntax. Something like:

```
Array<double, 1, 2> cross_ref = proj_ref_delta.asEigenMatrix().cross(proj_ref_ctr_delta.asEigenMatrix()/geom_dist_ref);
```

Reply:
So, I attempted something similar to this, however, it seems cross is only available for Eigen::Vector3d objects so I would have to double convert as there doesn't seem to be a direct ndarray->Eigen vector method. Options are: leave this as is or double convert the ndarrays from Matrix to Vector3d and store in the method for later use. What do ya think?

Reply:
Do you have a code example for this? Previously I had to dig through ndarray docs to find out to use asEigenMatrix properly.

Reply:
https://ndarray.github.io/ndarray/eigen_8h.html#ad1573ab63904bc25bde732c1f6161014

Reply:
Done. Worked with @TallJimbo on getting the cross product to work. The cross function has very specific needs in order to be used.